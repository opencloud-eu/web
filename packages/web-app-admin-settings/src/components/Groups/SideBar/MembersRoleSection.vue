<template>
  <ul class="oc-list">
    <li
      v-for="(member, index) in groupMembers"
      :key="index"
      class="oc-flex oc-flex-middle oc-mb-s"
      data-testid="group-members-list"
    >
      <user-avatar :user-id="member.id" :user-name="member.displayName" class="oc-mr-s" />
      <span class="oc-text-truncate" :title="member.displayName">
        {{ member.displayName }}
      </span>
    </li>
  </ul>
</template>
<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { User } from '@opencloud-eu/web-client/graph/generated'
import { UserAvatar } from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'MembersRoleSection',
  components: { UserAvatar },
  props: {
    groupMembers: {
      type: Array as PropType<User[]>,
      required: false,
      default: (): User[] => []
    }
  }
})
</script>
