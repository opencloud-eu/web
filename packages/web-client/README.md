# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
