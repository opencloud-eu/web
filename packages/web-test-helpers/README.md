# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
