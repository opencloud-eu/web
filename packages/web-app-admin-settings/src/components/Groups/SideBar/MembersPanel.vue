<template>
  <div class="oc-ml-s">
    <oc-text-input
      v-model="filterTerm"
      class="oc-text-truncate oc-mr-s oc-mt-m"
      :label="$gettext('Filter members')"
    />
    <div ref="membersListRef" data-testid="space-members">
      <div v-if="!filteredGroupMembers.length">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('No members found')" />
      </div>
      <div v-if="filteredGroupMembers.length" class="oc-mb-m" data-testid="group-members">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('Members')" />
        <members-role-section :group-members="filteredGroupMembers" />
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, inject, ref, watch, unref, Ref } from 'vue'
import MembersRoleSection from '../../Groups/SideBar/MembersRoleSection.vue'
import Fuse from 'fuse.js'
import Mark from 'mark.js'
import { Group, User } from '@opencloud-eu/web-client/graph/generated'
import { defaultFuseOptions } from '@opencloud-eu/web-pkg'

export default defineComponent({
  name: 'GroupsMembersPanel',
  components: { MembersRoleSection },
  setup() {
    const group = inject<Ref<Group>>('group')
    const filterTerm = ref('')
    const markInstance = ref(null)
    const membersListRef = ref(null)

    const filterMembers = (collection: User[], term: string) => {
      if (!(term || '').trim()) {
        return collection
      }

      const searchEngine = new Fuse(collection, { ...defaultFuseOptions, keys: ['displayName'] })
      return searchEngine.search(term).map((r) => r.item)
    }

    const members = computed(() => {
      if (group) {
        return unref(group).members.sort((a, b) => a.displayName.localeCompare(b.displayName))
      }
      return []
    })

    const filteredGroupMembers = computed(() => {
      return filterMembers(unref(members), unref(filterTerm))
    })

    watch(filterTerm, () => {
      if (unref(membersListRef)) {
        markInstance.value = new Mark(unref(membersListRef))
        unref(markInstance).unmark()
        const searchTermRegex = unref(filterTerm)
        unref(markInstance).mark(searchTermRegex, {
          element: 'span',
          className: 'mark-highlight'
        })
      }
    })

    return {
      filterTerm,
      membersListRef,
      members,
      filteredGroupMembers
    }
  }
})
</script>
