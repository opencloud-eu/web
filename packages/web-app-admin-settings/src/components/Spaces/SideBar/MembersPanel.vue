<template>
  <div class="oc-ml-s">
    <oc-text-input
      v-model="filterTerm"
      class="oc-mr-s oc-mt-m"
      :label="$gettext('Filter members')"
    />
    <div ref="membersListRef" data-testid="space-members">
      <div v-if="!filteredPermissions.length">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('No members found')" />
      </div>
      <div v-for="(role, i) in availableRoles" :key="i">
        <div
          v-if="getPermissionsForRole(role).length"
          class="oc-mb-m"
          :data-testid="`space-members-role-${role.displayName}`"
        >
          <h3 class="oc-text-bold oc-text-medium" v-text="role.displayName" />
          <members-role-section :permissions="getPermissionsForRole(role)" />
        </div>
      </div>
      <div v-if="permissionsWithoutRole.length" class="space-members-custom">
        <h3 class="oc-text-bold oc-text-medium" v-text="$gettext('Custom role')" />
        <members-role-section :permissions="permissionsWithoutRole" />
      </div>
    </div>
  </div>
</template>
<script setup lang="ts">
import { computed, inject, ref, watch, unref } from 'vue'
import { ShareRole, SpaceResource } from '@opencloud-eu/web-client'
import MembersRoleSection from './MembersRoleSection.vue'
import Fuse from 'fuse.js'
import Mark from 'mark.js'
import { defaultFuseOptions, useSharesStore } from '@opencloud-eu/web-pkg'
import { Permission } from '@opencloud-eu/web-client/graph/generated'

const sharesStore = useSharesStore()

const resource = inject<SpaceResource>('resource')
const filterTerm = ref('')
const markInstance = ref(null)
const membersListRef = ref(null)

const filterMembers = (collection: Permission[], term: string) => {
  if (!(term || '').trim()) {
    return collection
  }

  const searchEngine = new Fuse(collection, {
    ...defaultFuseOptions,
    keys: ['grantedToV2.user.displayName', 'grantedToV2.group.displayName']
  })
  return searchEngine.search(term).map((r) => r.item)
}

const permissions = computed(() => {
  return Object.values(unref(resource).root.permissions)
})

const filteredPermissions = computed(() => {
  return filterMembers(unref(permissions), unref(filterTerm))
})

const availableRoles = computed<ShareRole[]>(() => {
  const permissionsWithRole = unref(permissions).filter((p) => !!p.roles.length)
  const roleIds = [...new Set(permissionsWithRole.map((p) => p.roles).flat())]
  return roleIds
    .map((r) => sharesStore.graphRoles[r])
    .filter(Boolean)
    .sort((a, b) => {
      // sort roles by amount of permissions (most likely translates to manager > editor > viewer)
      const permissionsA = a.rolePermissions.flatMap((r) => r.allowedResourceActions)
      const permissionsB = b.rolePermissions.flatMap((r) => r.allowedResourceActions)
      return permissionsB.length - permissionsA.length
    })
})

const permissionsWithoutRole = computed(() => {
  return unref(filteredPermissions).filter(({ roles }) => !roles.length)
})

const getPermissionsForRole = (role: ShareRole) => {
  return unref(filteredPermissions).filter(({ roles }) => roles.includes(role.id))
}

watch(filterTerm, () => {
  if (unref(membersListRef)) {
    markInstance.value = new Mark(unref(membersListRef))
    unref(markInstance).unmark()
    unref(markInstance).mark(unref(filterTerm), {
      element: 'span',
      className: 'mark-highlight'
    })
  }
})
</script>
