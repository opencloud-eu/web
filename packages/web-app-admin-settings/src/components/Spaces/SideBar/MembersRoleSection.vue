<template>
  <ul class="oc-list">
    <li
      v-for="(m, index) in members"
      :key="index"
      class="oc-flex oc-flex-middle oc-mb-s"
      data-testid="space-members-list"
    >
      <oc-avatar
        v-if="m.grantedTo.user"
        :user-name="m.grantedTo.user.displayName"
        :width="36"
        class="oc-mr-s"
      /><oc-avatar-item
        v-else
        :width="36"
        icon-size="medium"
        :icon="groupIcon"
        name="group"
        class="oc-mr-s"
      />
      {{ (m.grantedTo.user || m.grantedTo.group).displayName }}
    </li>
  </ul>
</template>
<script lang="ts">
import { computed, defineComponent, PropType } from 'vue'
import { ShareTypes, SpaceMember } from '@opencloud-eu/web-client'

export default defineComponent({
  name: 'MembersRoleSection',
  props: {
    members: {
      type: Array as PropType<SpaceMember[]>,
      required: true
    }
  },
  setup() {
    const groupIcon = computed(() => {
      return ShareTypes.group.icon
    })
    return { groupIcon }
  }
})
</script>
