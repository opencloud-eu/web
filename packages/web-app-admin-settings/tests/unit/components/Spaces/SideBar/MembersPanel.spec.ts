import MembersPanel from '../../../../../src/components/Spaces/SideBar/MembersPanel.vue'
import { defaultPlugins, shallowMount } from '@opencloud-eu/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { ShareRole, SpaceResource } from '@opencloud-eu/web-client'
import MembersRoleSection from '../../../../../src/components/Spaces/SideBar/MembersRoleSection.vue'

const graphRoles = {
  '1': mock<ShareRole>({ id: '1', displayName: 'Managers', rolePermissions: [] }),
  '2': mock<ShareRole>({ id: '2', displayName: 'Editors', rolePermissions: [] }),
  '3': mock<ShareRole>({ id: '3', displayName: 'Viewers', rolePermissions: [] })
}

const spaceMock = {
  members: {
    '1': { roleId: '1', grantedTo: { user: { displayName: 'admin' } } },
    '2': { roleId: '2', grantedTo: { user: { displayName: 'marie' } } },
    '3': { roleId: '3', grantedTo: { user: { displayName: 'einstein' } } }
  }
} as undefined as SpaceResource

const selectors = {
  membersRolePanelStub: 'members-role-section-stub',
  spaceMembersCustom: '.space-members-custom'
}

describe('MembersPanel', () => {
  it('should render all members accordingly to their role assignments', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should filter members accordingly to the entered search term', async () => {
    const userToFilterFor = spaceMock.members['3']
    const { wrapper } = getWrapper()
    const input = wrapper.find('input')
    await input.setValue('ein')
    await wrapper.vm.$nextTick()
    expect(wrapper.findAll(selectors.membersRolePanelStub).length).toBe(1)
    expect(
      wrapper.findComponent<typeof MembersRoleSection>(selectors.membersRolePanelStub).props()
        .members[0].grantedTo.user.displayName
    ).toEqual(userToFilterFor.grantedTo.user.displayName)
  })
  it('should display an empty result if no matching members found', async () => {
    const { wrapper } = getWrapper()
    const input = wrapper.find('input')
    await input.setValue('no-match')
    await wrapper.vm.$nextTick()
    expect(wrapper.findAll(selectors.membersRolePanelStub).length).toBe(0)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should display members without role under the custom section', () => {
    const spaceResource = {
      members: {
        '1': { grantedTo: { user: { displayName: 'admin' } } }
      }
    } as undefined as SpaceResource
    const { wrapper } = getWrapper({ spaceResource })
    expect(wrapper.find(selectors.spaceMembersCustom).exists()).toBeTruthy()
  })
})

function getWrapper({ spaceResource = spaceMock } = {}) {
  return {
    wrapper: shallowMount(MembersPanel, {
      global: {
        stubs: { OcTextInput: false },
        plugins: [...defaultPlugins({ piniaOptions: { sharesState: { graphRoles } } })],
        provide: { resource: spaceResource }
      }
    })
  }
}
