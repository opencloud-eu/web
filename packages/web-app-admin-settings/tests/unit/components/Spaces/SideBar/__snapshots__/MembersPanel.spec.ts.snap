// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MembersPanel > should display an empty result if no matching members found 1`] = `
"<div class="oc-ml-s">
  <div class="oc-mr-s oc-mt-m"><label class="oc-label" for="oc-textinput-3">Filter members</label>
    <div class="oc-position-relative">
      <!--v-if--> <input id="oc-textinput-3" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text" value="no-match">
      <!--v-if-->
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div data-testid="space-members">
    <div>
      <h3 class="oc-text-bold oc-text-medium">No members found</h3>
    </div>
    <div>
      <!--v-if-->
    </div>
    <div>
      <!--v-if-->
    </div>
    <div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`MembersPanel > should render all members accordingly to their role assignments 1`] = `
"<div class="oc-ml-s">
  <div class="oc-mr-s oc-mt-m"><label class="oc-label" for="oc-textinput-1">Filter members</label>
    <div class="oc-position-relative">
      <!--v-if--> <input id="oc-textinput-1" aria-invalid="false" class="oc-text-input oc-input oc-rounded" type="text" value="">
      <!--v-if-->
    </div>
    <!--v-if-->
    <!--v-if-->
  </div>
  <div data-testid="space-members">
    <!--v-if-->
    <div>
      <div class="oc-mb-m" data-testid="space-members-role-Managers">
        <h3 class="oc-text-bold oc-text-medium">Managers</h3>
        <members-role-section-stub members="[object Object]"></members-role-section-stub>
      </div>
    </div>
    <div>
      <div class="oc-mb-m" data-testid="space-members-role-Editors">
        <h3 class="oc-text-bold oc-text-medium">Editors</h3>
        <members-role-section-stub members="[object Object]"></members-role-section-stub>
      </div>
    </div>
    <div>
      <div class="oc-mb-m" data-testid="space-members-role-Viewers">
        <h3 class="oc-text-bold oc-text-medium">Viewers</h3>
        <members-role-section-stub members="[object Object]"></members-role-section-stub>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
